mongo/mongo.go

```diff
@@ -228,6 +228,17 @@ func transformAggregatePipeline(registry *bsoncodec.Registry, pipeline interface
 		aidx, arr := bsoncore.AppendArrayStart(nil)
 		var hasOutputStage bool
 		valLen := val.Len()
+
+		// Explicitly forbid non-empty pipelines that are semantically single documents
+		// and are implemented as slices.
+		switch t := pipeline.(type) {
+		case bson.D, bson.Raw, bsoncore.Document:
+			if valLen > 0 {
```

Comment:
It's possible for `valLen` to be greater than zero but still represent an empty BSON document. For example, `bson.D{}` marshalled as BSON and represented as a `bson.Raw` is `bson.Raw{0x5, 0x0, 0x0, 0x0, 0x0}`. If a user passed in a `bson.D{}`, it would be allowed, but if they marshalled the same value as BSON and passed it as a `bson.Raw`, it would cause an error. Do we want to handle cases where someone is using an empty `bson.D{}` but marshalled as BSON?

Reply:
As it is currently, passing `bson.Raw{0x5, 0x0, 0x0, 0x0, 0x0}` as the `pipeline` parameter actually causes a `Marshal` error from `transformBsoncoreDocument` on L243 below. So, that use case is not supported. 

I _believe_ `tranformAggregatePipeline` is assuming the passed-in `pipeline` is an `interface{}` that must be `Marshal`ed to a `bsoncore.Document`. `bson.Raw{0x5, 0x0, 0x0, 0x0, 0x0}` has already been `Marshal`ed, so the second [`Marshal`](https://github.com/mongodb/mongo-go-driver/blob/master/mongo/mongo.go#L157) causes an error. Rather than support passing in an already marshaled pipeline, I think it's probably better to error with the "Use bson.A or mongo.Pipeline" message instead.

Reply:
Are there any `bson.Raw` or `bsoncore.Document` values that  `transformAggregatePipeline()` supports? If not, we don't need to check `valLen` for those types, just return an error for any `bson.Raw` or `bsoncore.Document` (or `[]byte`?) value.

e.g.
```go
switch t := pipeline.(type) {
case bson.D:
	if len(t) > 0 {
		return fmt.Errorf(...)
	}
case bson.Raw, bsoncore.Document:
	return fmt.Errorf(...)
}
```

Reply:
Empty `bson.Raw` and `bsoncore.Document` instances are supported by `transformAggregatePipeline()`. I've added a test case for each to ensure they don't error.

Reply:
Ah, thanks for adding those test cases. Sounds good 👍